cli/src/cmd/deploy.rs

```diff
@@ -0,0 +1,172 @@
+// along with ink!.  If not, see <http://www.gnu.org/licenses/>.
+
+use crate::{
+    cmd::{
+        build,
+        Result,
+    }
+};
+
+use futures::future::Future;
+use runtime_primitives::{
+    generic::{Era, Header},
+    traits::{IdentityLookup, Verify, BlakeTwo256},
+    AnySignature,
+};
+use std::{
+    collections::HashMap,
+    fs,
+    io::Read,
+    path::{Path, PathBuf},
+};
+use substrate_primitives::{
+    crypto::Pair,
+    sr25519,
+    H256,
+};
+use subxt::{
+    balances::Balances,
+    contracts::{
+        Contracts,
+        ContractsXt,
+    },
+    system::System,
+};
+
+/// Load the wasm blob from the specified path, or defaults to the target contract wasm in the
+/// current project, inferred via the crate metadata.
+fn load_contract_code(path: Option<&PathBuf>) -> Result<Vec<u8>> {
+    let default_wasm_path = build::collect_crate_metadata()
+        .map(|metadata| metadata.dest_wasm())?;
+    let contract_wasm_path = path.unwrap_or(&default_wasm_path);
+
+    let mut data = Vec::new();
+    let mut file = fs::File::open(&contract_wasm_path)
+        .map_err(|e| format!("Failed to open {}: {}", contract_wasm_path.display(), e))?;
+    file.read_to_end(&mut data)?;
+
+    return Ok(data)
+}
+
+/// If the extrinsic was successfully included in a block, attempt to extract the code hash from the
+/// `Contracts::CodeStored` event
+fn extract_code_hash(extrinsic_result: subxt::ExtrinsicSuccess<Runtime>) -> Result<H256> {
+    match extrinsic_result.find_event::<H256>("Contracts", "CodeStored") {
+        Some(Ok(hash)) => Ok(hash),
+        Some(Err(err)) => Err(format!("Failed to decode code hash: {}", err).into()),
+        None => Err("Failed to find Contracts::CodeStored Event".into()),
+    }
+}
+
+/// Define a Runtime to provide a subset of types for interacting with the target chain via RPC.
+///
+/// # Note
+///
+/// The concrete types MUST be compatible with the target chain's respective concrete type
+/// definitions.
+#[derive(Debug, Clone, Eq, PartialEq)]
+struct Runtime;
+
+impl System for Runtime {
+    type Index = u32;
+    type BlockNumber = u32;
+    type Hash = substrate_primitives::H256;
+    type Hashing = BlakeTwo256;
+    type AccountId = <AnySignature as Verify>::Signer;
+    type Address = srml_indices::address::Address<Self::AccountId, u32>;
+    type Header = Header<Self::BlockNumber, BlakeTwo256>;
+}
+
+impl Balances for Runtime {
+    type Balance = u64;
+}
+
+impl Contracts for Runtime {}
+
+/// Put contract code to a smart contract enabled substrate chain.
+/// Returns the code hash of the deployed contract if successful.
+///
+/// Optionally supply the contract wasm path, defaults to destination contract file inferred from
+/// Cargo.toml of the current contract project.
+///
+/// Creates an extrinsic with the `Contracts::put_code` Call, submits via RPC, then waits for
+/// the `ContractsEvent::CodeStored` event.
+pub(crate) fn execute_deploy(
+    url: url::Url,
+    surl: &str,
```

Comment:
`surl` ? is that a thing? :upside_down_face:

Reply:
Should be `suri` (secret uri). Copied from `subkey` https://github.com/paritytech/substrate/blob/5a2be8bd56ed5d9dc787fb3b9a76ce10ced4704b/subkey/src/main.rs#L234